addons/stock/models/stock_location.py

```diff
@@ -187,6 +201,63 @@ def should_bypass_reservation(self):
         self.ensure_one()
         return self.usage in ('supplier', 'customer', 'inventory', 'production') or self.scrap_location or (self.usage == 'transit' and not self.company_id)
 
+    def _check_can_be_used(self, product, quantity=0, package=None):
+        """Check if product/package can be stored in the location. Quantity
+        should in the default uom of product, it's only used when no package is
+        specified."""
+        self.ensure_one()
+        if package and package.packaging_id:
+            return self._check_package_storage(product, package=package)
+        return self._check_product_storage(product, quantity=quantity)
+
+    def _check_product_storage(self, product, quantity):
+        """Check if a number of product can be stored in the location. Quantity
+        should in the default uom of product."""
+        self.ensure_one()
+        if self.storage_category_id:
+            # check weight
+            if self.storage_category_id.max_weight < product.weight:
+                return False
+            # check if only allow new product when empty
+            if self.storage_category_id.allow_new_product =="empty" and self.quant_ids:
+                return False
+            # check if only allow same product
+            if self.storage_category_id.allow_new_product =="same" and self.quant_ids and self.quant_ids.product_id != product:
+                return False
+            # check if enough space
+            product_capacity = self.storage_category_id.product_capacity_ids.filtered(lambda pc: pc.product_id == product)
+            if product_capacity:
+                product_qty = sum(self.quant_ids.filtered(lambda q: q.product_id == product).mapped('quantity'))
+                reserved_qty = sum(self.env['stock.move.line'].search([
+                    ('product_id', '=', product.id),
+                    ('location_dest_id', '=', self.id),
+                    ('state', 'not in', ['draft', 'done', 'cancel'])
+                ]).mapped('product_qty'))
+                if product_qty + reserved_qty + quantity > product_capacity.quantity:
+                    return False
+        return True
+
+    def _check_package_storage(self, product, package):
+        """Check if the given package can be stored in the location."""
+        self.ensure_one()
+        if self.storage_category_id:
+            # check weight
+            if self.storage_category_id.max_weight < package.packaging_id.max_weight:
+                return False
+            # check if only allow new product when empty
+            if self.storage_category_id.allow_new_product =="empty" and self.quant_ids:
+                return False
+            # check if only allow same product
```

Comment:
when receiving lots, I see several cases
- each reception has a different lot, and you always receive a unique lot -> just use 'empty only' in the storage category
- each reception can have several lots, but in this case I assume you  don't receive a lot that you already have in your stock. if you want to use one lot per location, just do a put in pack at reception and send to a 'empty only' storage category 
So I don't think it would so useful to add this lot dimension, sounds like a step too far to me but I may be wrong.